### ISSUE TITLE
Destroying component doesn't remove references to DOM elements and keeps them as detached nodes in the memory. 

### ISSUE THREAD

**Issue by AtanasChachev** [ionic-team/ionic-v3#1040]
**I'm submitting a ...**  (check one with "x")
[ x ] bug report
[ ] feature request

**Current behavior:**

After playing with an application for some time, the DOM Nodes in memory stacks when navigating between components.

**Expected behavior:**

DOM Nodes and their references to be destroyed and released from the memory when the component is destroyed.

Hi guys. We have built an app in ionic v3 with angular 5 , the issue we have is that when playing with app after some time it slows down on slow machines (Under slow machines I mean Kiosk machines with low specifications ). After some profiling and taking snapshots, it turned out that DOM Nodes are stacking up and are not cleared when the component is destroyed (e.g from Component1 navigating to Component2 -> After snapshot is taken , there are Detached nodes (Detached HTMLElements and etc.) from Component 1). 

At first we thought that the issues are coming from the Modal Controller ( because we do have a lot of modals in our app ) and we've updated to ionic-angular 3.9.6 - related to https://github.com/ionic-team/ionic/issues/12574. But unfortunately that didn't solve our issue.

**We do use Electron for desktop app, on iOS we don't face that issue , but on Electron we do since it runs with headless chrome browser. Also this issue is reproducable after 20-30 minutes playing with the app. At first it runs smooth.**

The most confusing part is  that when we have the following element:

`<div tappable (click)="functionX()" (press)="functionY()">Execute function</div>`

after the slow down, on click the both functions (click) and (press) are executed at the same time.

Unfortunately cannot share code because the app is massive but gonna mention the things we do.

1. We do use SSE (ServerSent events)
2. We do use @ViewChild() to component references with `{ read: ElementRef }`
3. We do use @Hostlisteners in the components
4. For the requests we'r using Promises
5. There are few of subscriptions but they'r unsubscribed (anyway we have 2 subscriptions in the app.component.ts) only
6. We do use `trackBy` for `ngFor`
7. We do use *ngIf with `<ng-container></ng-container>` (e.g we don't have `<span *ngIf="isTrue">Render</span>`)
8. We do use virtual scroll for the lists.

Sharing screenshots from the profiling:

<img width="1278" alt="Screenshot_2" src="https://user-images.githubusercontent.com/10319093/60418886-bb642380-9bec-11e9-9789-2a5117ed0a24.png">

![Screenshot_1](https://user-images.githubusercontent.com/10319093/60418860-a7202680-9bec-11e9-9e9e-08fac62d32b4.png)

**The results are achieved when running the app with `ionic serve` locally under CPU slowdown 6x in the browser to reproduce the issue**

Sharing the dependencies:

> "dependencies": {
        "@angular/animations": "5.2.11",
        "@angular/common": "5.2.11",
        "@angular/compiler": "5.2.11",
        "@angular/compiler-cli": "5.2.11",
        "@angular/core": "5.2.11",
        "@angular/forms": "5.2.11",
        "@angular/http": "5.2.11",
        "@angular/platform-browser": "5.2.11",
        "@angular/platform-browser-dynamic": "5.2.11",
        "@ionic-native/background-mode": "^4.18.0",
        "@ionic-native/core": "~4.11.0",
        "@ionic-native/device": "^4.12.2",
        "@ionic-native/device-motion": "^4.17.0",
        "@ionic-native/in-app-browser": "^4.18.0",
        "@ionic-native/screen-orientation": "^4.12.0",
        "@ionic-native/splash-screen": "~4.11.0",
        "@ionic-native/status-bar": "~4.11.0",
        "@ionic/storage": "2.1.3",
        "@ngx-translate/core": "^9.1.1",
        "@ngx-translate/http-loader": "^3.0.1",
        "cordova-ios": "4.5.5",
        "cordova-ios-plugin-no-export-compliance": "0.0.5",
        "cordova-plugin-background-mode-fixed": "^0.7.6",
        "cordova-plugin-device": "^2.0.2",
        "cordova-plugin-device-motion": "^2.0.1",
        "cordova-plugin-inappbrowser": "^3.0.0",
        "cordova-plugin-ionic-keyboard": "^2.1.3",
        "cordova-plugin-ionic-webview": "^2.4.1",
        "cordova-plugin-screen-orientation": "^3.0.1",
        "cordova-plugin-splashscreen": "^5.0.2",
        "cordova-plugin-whitelist": "^1.3.3",
        "cordova-sqlite-evcore-extbuild-free": "^0.9.10",
        "cordova-windows": "6.0.1",
        "electron-log": "^2.2.17",
        "electron-updater": "^4.0.6",
        "event-source-polyfill": "^1.0.5",
        "http-server": "^0.11.1",
        "ionic-angular": "3.9.6",
        "ionicons": "3.0.0",
        "moment": "2.24.0",
        "ng-inline-svg": "^8.0.2",
        "ng2-datepicker": "^2.3.1",
        "ngx-electron": "^2.1.1",
        "ngx-pinch-zoom": "^1.2.0",
        "ngx-swiper-wrapper": "^7.2.1",
        "ngx-virtual-scroller": "^1.0.17",
        "rxjs": "5.5.11",
        "sw-toolbox": "3.6.0",
        "zone.js": "0.8.26"
    }

**Comment by longgt** [comment]
https://github.com/ionic-team/ionic/issues/12574#issuecomment-342344604

ionic 3.9.6 only fixed `orientationchange` problem, not `componentRef` as described above.

Without reproduce repo, it's hard to do some investigation more. (A repo with only Component1, Component2 is OK)

**Comment by AtanasChachev** [comment]
@longgt I will try to create some sample for this one. By the way the DOM Nodes that are Detached are  component elements, let's say span , div etc...

**Comment by AtanasChachev** [comment]
Hey @longgt thanks for you answer. I've just tried an ionic 4 scratch app and just toggling between the 2 generated pages from the menu causing the same thing. 

Basically what I did is

**Steps to reproduce:** 

1.ionic start testApp --type=angular with `sidebar` 

2.then changed the structure in app.component.html to

`<ion-app>

  <ion-menu side="start" menuId="first">

    <ion-header>
      <ion-toolbar>
        <ion-title>Menu</ion-title>
      </ion-toolbar>
    </ion-header>
    <ion-content>
      <ion-list>
        <ion-menu-toggle auto-hide="false" *ngFor="let p of appPages">
          <ion-item [routerDirection]="'root'" [routerLink]="[p.url]">
            <ion-icon slot="start" [name]="p.icon"></ion-icon>
            <ion-label>
              {{p.title}}
            </ion-label>
          </ion-item>
        </ion-menu-toggle>
      </ion-list>
    </ion-content>
  </ion-menu>
    <ion-router-outlet main></ion-router-outlet>
</ion-app>
`

3. Do the instructions from above
4. Just open the side menu , enter Home or List page , redo that 10 , 20 times while profiling , then dont do anything... the nodes are stacking in memory


**NOTE: This also happened with ionic v3 scratch created app**

and then got this result 

![Screenshot 2019-07-03 at 21 23 17](https://user-images.githubusercontent.com/10319093/60616373-f8413d80-9dd9-11e9-9a34-94be125f7139.png)


Anyway, I've tested pure Angular 5 app that we've developed into the past, which is super massive , when ran profiling, a decrease of nodes was noticed.

![Screenshot 2019-07-03 at 21 28 00](https://user-images.githubusercontent.com/10319093/60616429-20c93780-9dda-11e9-9b6e-dde426ce3f1a.png)

**Comment by longgt** [comment]
@AtanasChachev maybe related to https://github.com/angular/angular/issues/22155

**Comment by AtanasChachev** [comment]
@longgt not quite sure this is the problem the screenshot from above is captured from the ionic v4 project, I've also checked the `package.json` on the ionic v4 it uses

`"@angular/platform-browser": "^7.2.2",`
` "@angular/platform-browser-dynamic": "^7.2.2",`

Also checked the code 

https://github.com/JiaLiPassion/angular/commit/33ce0da746e6ceb2813c723c46ac714239d72cd7

this one is implemented in the `hammer_gestures.js` file :(

**Comment by longgt** [comment]
@AtanasChachev this is ionic/v3 issues tracking. Your project is using angular 5 so the bug fixing of 22155 is not patched yet.

If you have same issue with ionic/v4, fire a ticket to v4 repo instead.

**Comment by longgt** [comment]
Workaround here is trying to create your own `HammerGestureConfig` which extends from `IonicGestureConfig`, then apply the fix of 22155 PR to it.

**Comment by AtanasChachev** [comment]
@longgt Tried the fix on the project that the issue has been reported for.

**Result: Nodes are significantly decreasing when the component is destroyed.**

Anyway - unfortunately the `addEventListener` prototype is in the `HammerGesturesPlugin` and not `HammerGestureConfig` and cannot be extended in provider and be overwritten as GustureConfig can... so for the test I did it through node_modules/@angular/browser-platform/esm5/platform-browser.js.

**Comment by longgt** [comment]
@AtanasChachev We can add some monkey patching to object which returned from `buildHammer` function like below

```
@Injectable()
export class CustomGestureConfig extends HammerGestureConfig {
    overrides: {[key: string]: Object} = {
        pinch: {enable: false},
        rotate: {enable: false}
    };

    /**
     * Build hammer from HTMLElement. Returns proxy object overriding on/off method.
     * @param element 
     */
    buildHammer( element: HTMLElement ) {
        let mc = new (<any> window).Hammer(element);

        for (let eventName in this.overrides) {
          mc.get(eventName).set(this.overrides[eventName]);
        }

        return {
            on: ( eventName: string, callback: Function ) => {
                return mc.on( eventName, callback );
            },
            off: ( eventName: string, callback: Function ) => {
                const omc = mc.off( eventName, callback );
                //Manually calling destroy
                if (typeof omc.destroy === 'function') {
                  omc.destroy();
                }
                
                return omc;
            }
        };
    }
}
```

then provide it as default `HammerGestureConfig`

app.module.ts
```
{ provide: HAMMER_GESTURE_CONFIG, useClass: CustomGestureConfig },
```

**Comment by AtanasChachev** [comment]
Works like a charm. Thanks a lot @longgt !! I'm closing this issue as resolved since after this solution decreased nodes in the memory significantly. 

**NOTE: Since we also do use <ion-digit-keyboard> component in our project , since we have decreased the nodes , there were still some nodes left in the memory and that <ion-digit-keyboard> were causing them. Anyway we have managed to fix this too**

I'm closing this issue. Again thanks @longgt , cheers.
